### ISSUE TITLE
Validation context is not usable via the model constructor

### ISSUE THREAD

**Issue by jason-wolfe** [pydantic/pydantic#5623]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

As part of pydantic V2, #1549 introduced the ability to access context external to model fields when validating data coming through `model_validate` and `model_validate_json`. This is achieved via passing in an extra named parameter `context`, which is available to validators at `info.context`.

Using these models through the model's constructor does not work:
1. major: it is not possible to set `info.context` via the model's `__init__` 
2. minor: `info.context` has the value `None` when using the model's `__init__`, instead of being an empty dict. This makes defaulting hard to get right.

## A demo of how code might look when using the feature

```python3
class MyModel(BaseModel):
    x: int
    
    @field_validator("x")
    def validate(cls, value, info):
        return value * info.context.get("multiplier", 1)

# works great
assert MyModel.model_validate({"x": 1}).x == 1
assert MyModel.model_validate({"x": 1}, context={"multiplier": 2}).x == 2

# this line should work, but currently fails validation
model = MyModel(x=1) 
assert model.x == 1

model = MyModel(x=1, context={"multiplier": 2})
assert model.x == 2
```

I would not mind whatsoever if `context` got changed to `_context`, `_validation_context`, or whatever.

## Your use case(s) for the feature

Use-cases include:
* Having a dynamic set of allowed values
* Having validation strictness be controlled by a runtime flag
* See also the list in: https://github.com/pydantic/pydantic/issues/1549

My personal immediate use-case is controlling allowed values dynamically.

## Why the feature should be added to pydantic (as opposed to another library or just implemented in your code)

Without a way to use context via the constructor, pydantic only partially supports validation context. I think that this is enough justification for bringing a solution into pydantic itself.

## Alternatives

1. Just forget about context altogether, and pass the validation context as part of the model. This should be ruled out because it makes it very easy to create injection attacks whereby the end-user can control validation context by passing in a malicious payload.
2. Pass context via a ContextVar, as demonstrated [here](https://github.com/pydantic/pydantic/issues/1549#issuecomment-1526632046). This is workable, but requires calling into pydantic-internal code directly. If it is the preferred solution, it belongs as a base class / context manager in the library in order to round out validation context usability.
3. Override the constructor to take a positional argument, as also demonstrated [here](https://github.com/pydantic/pydantic/issues/1549#issuecomment-1526632046). This is workable, but thwarts static type-checking if done externally. If this becomes a part of pydantic, then (I assume) it can also be a part of the mypy plugin, giving proper type-checking.

(Thanks @dmontagu for both of the viable alternatives!)

### Affected Components

- [X] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [X] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by dmontagu** [comment]
I think if we were going to make it possible to pass the context to `__init__`, we'd need it to be a keyword argument starting with `__` to prevent collisions with aliased field names.

Also, unfortunately, I think this would not play nice with type checkers that make use of `@dataclass_transform`.

I did make a proposal of a way to pass context into `__init__` _without_ adding any keyword arguments through the use of a `ContextVar` + `contextmanager` in https://github.com/pydantic/pydantic/issues/1549#issuecomment-1526632046. This has the upshot of requiring no changes today, but does require the user to override the `BaseModle.__init__` function. Fortunately, the (current) behavior of overriding `BaseModel.__init__` does exactly what you'd want here, since `__init__` is only called if you instantiate the class from Python, but it could lead to issues if we rework how custom `__init__` behaves.

@adriangb @samuelcolvin, I'd appreciate your thoughts on this.

**Comment by jason-wolfe** [comment]
> I did make a proposal of a way to pass context into `__init__` _without_ adding any keyword arguments through the use of a `ContextVar` + `contextmanager` in [#1549 (comment)](https://github.com/pydantic/pydantic/issues/1549#issuecomment-1526632046). This has the upshot of requiring no changes today, but does require the user to override the `BaseModle.__init__` function.

Yep! Thanks for the suggestion. I referenced that in alternative #2. I think it's a totally valid way to go, and if taken then should be added to the library itself, in order to complete the validation context functionality in pydantic.

**Comment by dmontagu** [comment]
Oh sorry, I didn't read your issue closely enough, and see the alternatives listed there. (Thanks for providing all the detail)

**Comment by dmontagu** [comment]
I'll note that the approach using a `ContextVar` is now officially documented: https://github.com/pydantic/pydantic/blob/6bffa13615ac4b91fa30d6818416f62152944dbe/docs/usage/validators.md?plain=1#L600

I think the main problem I have with having an always-working out-of-the-box solution for this is that, the obvious ways I can see to get that working (such as modifying the `BaseModel.__init__` to look for a specific library-provided `ContextVar`) involve adding overhead for the scenario where this functionality isn't used, and currently I think this will be relevant to only a small fraction of calls to model class `__init__` calls.

I personally would be open to adding a library-provided `ContextVar` and paired `@contextmanager`, with a bit of documentation (I think it would be enough for it to be in the docstrings of those functions and with minor tweaks to the docs section linked above) of how to use both together with a custom `__init__` to achieve this with a bit less user code.

However, given it's not clear to me how high of a priority this is though, I'll hold off on doing this myself until there is more of a widespread request for this. I would be happy to review a PR adding it though.

**Comment by jason-wolfe** [comment]
Sounds sensible to me, thanks for the write-up!

**Comment by HarishHary** [comment]
Hello,

Also interested in this feature. Need to pass-around data from parent model to child model and validate depending on context. Would be nice to initiate it via the __init__ function?

Is there a way to set the context during validation or using a special function?

**Comment by dmontagu** [comment]
@HarishHary this is documented here: https://docs.pydantic.dev/latest/usage/validators/#using-validation-context-with-basemodel-initialization

**Comment by aantonw** [comment]
Hi, I have same problem with nested model, where child needed to access parent value. I've tried to override `Parent` constructor and also `model_validator(mode="wrap")` but with no luck. Both called after `Child` validator run.


```python
from pydantic import BaseModel, ValidatorInfo, field_validator, conint

allowed_choices = ("a", "b", "c", "d", "e")


class Child(BaseModel):
    choice: str  # need to validate choice

    @field_validator("choice")
    def validate_choice(v, info: ValidatorInfo):
        # _choices = allowed_choices[:Parent.options]  ### how to access Parent.options here?
        _choices = allowed_choices
        if v not in _choices:
            raise ValueError(f"choice must be one of {_choices}")
        return v


class Parent(BaseModel):
    options: conint(ge=1, le=5)
```

**Comment by nextmat** [comment]
@dmontagu - Thank you for the `__init__` workaround, very helpful.

I'm working with a case where I want to vary the validation model for a child attribute based on context, for example:

```py
class Animal(BaseModel):
    name: str

class Dog(Animal):
    # dog-specific fields...

class Cat(Animal):
    # cat-specific fields...

class Person(BaseModel):
    name: str
    pet: Cat | Dog
 
person_args = {
    "name": "Dan",
    "pet": { 
        "name": "Fido",
        # dog-specific fields...
    },
} 

# ideally 
person = Person(**person_args, context={"pet_type": "dog"})

# or
with init_context({"pet_type": "dog"}):
    person = Person(**person_args)
```

From examples I can figure out how to do custom validation within a type (within a `str` or `list`), but I'm struggling with how to actually set the desired model for validating the child attribute (`pet` in this case). 

Is there a way to set the right member of the union at initialization using the context?
